---
title: "Transcriptome analysis of the `r params$tissue`-biopsies from the QTL family cohort"
author: 
- "Analysis by Thomas Gade Koefoed"
- "Research assistant, Hansen Group"
- "Novo Nordisk Foundation Center For Basic Metabolic Research"
- "thomas.koefoed@sund.ku.dk"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
   html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
    df_print: kable
params:
  genome: human
  genes: NCBI
  tissue: muscle
  exclude: none
  report_name: analysis
editor_options:
  chunk_output_type: inline
---
# Pre-processing {.tabset}
## Setup
```{r setup, warning=FALSE, message=FALSE, results="hide"}
library("tidyverse") # general data manipulation

script_name <- rstudioapi::getSourceEditorContext() %>% 
  `[[`("path") %>% 
  basename() %>% 
  tools::file_path_sans_ext()

date <- format(Sys.time(), "%Y-%m-%d-%H.%M")

data_dir <- stringr::str_glue("{params$report_name}_data/") %>%  
  as.character()
dir.create(data_dir, showWarnings = F)

suppressPackageStartupMessages({
  # library(conflicted)
  library('biomaRt') # downloading gene data
  library("DT") # for pretty printing of paged tables
  library("edgeR") # differential expression analysis
  library("here") # Better structure and relative paths
  library("GO.db") # downloading gene data
  
  if(params$genome == "human") {library("org.Hs.eg.db")
  } else if (params$genome == "mouse") {library("org.Mm.eg.db")
  } else if (params$genome == "rat") {library("org.Rn.eg.db")}
  
  library("AnnotationDbi")
  library("ggrepel") 
  library("here") # better path management
  library("cbmr") # in-house convenience functions
  library("plotly") # interactive plots
  # library("familyStudyPackage") # own package
})
i_am("DESCRIPTION")
devtools::load_all(here())
theme_set(theme_bw())
```

## Project specific data loading and formattings {.tabset}
### Create metadata framework from Salmon data
```{r}
combined_pheno_data <- get_combined_pheno_data()
combined_pheno_data_single_tissue <- combined_pheno_data %>% 
  filter(Tissue_fixed == params$tissue) %>% 
  filter(sourceId!="641-9") #%>% 
# filter(WH_risk=="High")
```

## DGE list creation for all samples

```{r, message = FALSE, class.source = 'fold-show'}
rna_counts_unordered <- get_rna_counts()
rna_counts <- rna_counts_unordered[,as.character(combined_pheno_data_single_tissue$Salmon.ID.raw)]
y_all <- DGEList(rna_counts, 
                 samples=combined_pheno_data_single_tissue)

log_cpm_all <- cpm(y_all, log=T) 
log_cpm_df_all <- log_cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

cpm_all <- cpm(y_all) 
cpm_df_all <- cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

design_all <-  model.matrix(~BIGG_SI_t_0_60_120 + SEX + AGE, 
                            data=y_all$samples, 
                            na.action="na.pass")
colnames(design_all) <- make.names(colnames(design_all))

idx_expressed_all <- filterByExpr(y_all, design = design_all) 


y_all <- y_all[idx_expressed_all,] %>% 
  calcNormFactors() %>%
  estimateDisp(design=design_all)
# subset by expressed genes and samples with the metadata
# y_all <- y_all[idx_expressed_all, !is.na(combined_pheno_data_single_tissue[["bmi"]])] %>% 
```

## Get gene info and print

```{r, class.source = 'fold-show'}
# download new database
# ensembl <- useMart("ensembl")
# ensembl <-  useDataset("rnorvegicus_gene_ensembl",mart=ensembl)
# saveRDS(ensembl, here("shared_data/rat_mart.RDS"))

# Get Description, external gene name via Ensembl/biomaRt
mart <-  readRDS(file = if (params$genome=="mouse") here("shared_data/mouse_mart.RDS")
                 else if (params$genome=="human") here("shared_data/human_mart.RDS")
                 else if (params$genome=="rat") here("shared_data/rat_mart.RDS"))


gene_info_list<- getBM(filters= case_when(params$genes=="NCBI" ~"entrezgene_accession",
                                          params$genes=="ENSEMBL" ~ "ensembl_gene_id"), 
                       attributes= c("entrezgene_accession", "ensembl_gene_id","description"),
                       values=rownames(y_all),
                       mart=mart) %>% 
  dplyr::rename("Name"=entrezgene_accession, 
                "ENSEMBL_ID"=ensembl_gene_id)

write_tsv(x = log_cpm_df_all, file=paste0(data_dir, "log_CPM_values_by_sample.tsv"))
```

## Session info

```{r}
sessionInfo()
```

# Analysis Code {.tabset}

## EdgeR all

```{r}
fit_all <- glmQLFit(y_all, design = design_all, robust = TRUE)

coefficients <- design_all %>% 
  colnames() %>% 
  `[`(-1) %>% 
  set_names()

DE_genes_test_all <- coefficients %>% 
  purrr::map(~edgeR_tester(coef = .x, 
                           efit = fit_all, 
                           id_name = "Name"))

DE_genes_test_all_w_info <- DE_genes_test_all %>% 
  map(~dplyr::left_join(., y=gene_info_list))
```

## GO all

```{r, message=FALSE, class.source = 'fold-show'}
# options <- expand_grid(c("human","mouse","rat"), c("SYMBOL", "ENSEMBL"))
# 
# pmap(options, ~{
#   GO_terms_all <- get_enrichment_terms(org_db =if (..1 == "human") {org.Hs.eg.db
#   } else if (..1 =="mouse") {org.Mm.eg.db
#   } else if (..1 =="rat") {org.Rn.eg.db},
#   gene_ids = keys(if (..1 == "human") {org.Hs.eg.db
#   } else if (..1 =="mouse") {org.Mm.eg.db
#   } else if (..1 =="rat") {org.Rn.eg.db}, keytype = ..2), 
#   gene_id_key_type = ..2,
#   min_genes = 5, max_genes = 500)$BP
#   
#   saveRDS(GO_terms_all, str_glue("shared_data/GO_terms_BP_{..1}_{..2}.RDS"))
# }
# )

GO_keytype <- case_when(params$genes=="ENSEMBL" ~ "ENSEMBL",
                       params$genes=="NCBI" ~ "SYMBOL")

GO_terms_all <- readRDS(str_glue("shared_data/GO_terms_BP_{params$genome}_{GO_keytype}.RDS"))

GO_terms_all <- get_enrichment_terms(org_db =org.Hs.eg.db,
  gene_ids = rownames(y_all),
  gene_id_key_type = "SYMBOL",
  min_genes = 5, max_genes = 500)$BP

# not equivalent?!
ontology_tests_all <- run_ontology_tests(GO_terms_all, 
                                         y_all, 
                                         coefs = coefficients,
                                         fun=limma::camera)

# COPY STUFF FROM 0193 here
```

## Save R-files
```{r}
saveRDS(y_all, file.path(data_dir,"y_all.RDS"))
saveRDS(DE_genes_test_all_w_info, file.path(data_dir,"DE_genes_test_all_w_info.RDS"))
saveRDS(ontology_tests_all, file.path(data_dir,"ontology_tests_all.RDS"))
saveRDS(fit_all, file.path(data_dir,"fit_all.RDS"))
```


# Results

## Summary
```{r, warning=F}
coefficient_count <- 1
summary_table <- map(list("Differentially expressed genes (FDR < 5%)"=DE_genes_test_all_w_info, 
                          "Enriched GO-terms (FDR < 5%)"=ontology_tests_all), 
                     n_sig_genes_pr_contrast) %>% 
  bind_rows(.id="Analysis") %>% knitr::kable()
summary_table
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated(only_sig=F)
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated()
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated()
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated()
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated()
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable_updated()
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable_updated()
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated()
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
coefficient_count <- coefficient_count+1
if (coefficient_count > length(coefficients)) {knitr::knit_exit()}
```


