---
title: "Transcriptome analysis of the `r params$tissue`-biopsies from the QTL family cohort"
author: 
- "Analysis by Thomas Gade Koefoed"
- "Research assistant, Hansen Group"
- "Novo Nordisk Foundation Center For Basic Metabolic Research"
- "thomas.koefoed@sund.ku.dk"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
   html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
    df_print: kable
params:
  genome: human
  genes: NCBI
  tissue: muscle
editor_options:
  chunk_output_type: inline
---
# Pre-processing {.tabset}
## Setup
```{r setup, warning=FALSE, message=FALSE, results="hide"}
full_script_name <- rstudioapi::getSourceEditorContext()$path
script_name <- basename(full_script_name)
data_dir <- paste0(tools::file_path_sans_ext(full_script_name),"_data/")
dir.create(data_dir, showWarnings = F)

suppressPackageStartupMessages({
  # library(conflicted)
  library('biomaRt') # downloading gene data
  library("DT") # for pretty printing of paged tables
  library("edgeR") # differential expression analysis
  library("here") # Better structure and relative paths
  library("GO.db") # downloading gene data
  
  if(params$genome == "human") {library("org.Hs.eg.db")
  } else if (params$genome == "mouse") {library("org.Mm.eg.db")
  } else if (params$genome == "rat") {library("org.Rn.eg.db")}
  
  library("AnnotationDbi")
  library("ggrepel") 
  library("here") # better path management
  library("cbmr") # in-house convenience functions
  library("tidyverse") # general data manipulation
  library("plotly") # interactive plots
  # library("familyStudyPackage") # own package
})
i_am("DESCRIPTION")
devtools::load_all(here())
knitr::opts_chunk$set(cache=F)
theme_set(theme_bw())
```

## Project specific data loading and formattings {.tabset}
### Create metadata framework from Salmon data
```{r}
combined_pheno_data <- get_combined_pheno_data()
combined_pheno_data_single_tissue <- combined_pheno_data %>% 
  filter(Tissue_fixed == params$tissue) %>% 
  filter(sourceId!="641-9")

head(combined_pheno_data_single_tissue)
```

## DGE list creation for all samples

```{r, message = FALSE, class.source = 'fold-show'}
rna_counts_unordered <- get_rna_counts()
rna_counts <- rna_counts_unordered[,as.character(combined_pheno_data_single_tissue$Salmon.ID.raw)]
y_all <- DGEList(rna_counts, 
                 samples=combined_pheno_data_single_tissue)

log_cpm_all <- cpm(y_all, log=T) 
log_cpm_df_all <- log_cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

cpm_all <- cpm(y_all) 
cpm_df_all <- cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

design_all <-  model.matrix(~BMI + smoker + SEX + AGE + HOMA_B_2.88 + HOMA_IR_156, 
                            data=y_all$samples, 
                            na.action="na.pass")

idx_expressed_all <- filterByExpr(y_all, design = design_all) 


y_all <- y_all[idx_expressed_all,] %>% 
  calcNormFactors() %>%
  estimateDisp(design=design_all)
# subset by expressed genes and samples with the metadata
# y_all <- y_all[idx_expressed_all, !is.na(combined_pheno_data_single_tissue[["bmi"]])] %>% 
```

## Get gene info and print

```{r, class.source = 'fold-show'}
# download new database
# ensembl <- useMart("ensembl")
# ensembl <-  useDataset("rnorvegicus_gene_ensembl",mart=ensembl)
# saveRDS(ensembl, here("shared_data/rat_mart.RDS"))

# Get Description, external gene name via Ensembl/biomaRt
mart <-  readRDS(file = if (params$genome=="mouse") here("shared_data/mouse_mart.RDS")
                 else if (params$genome=="human") here("shared_data/human_mart.RDS")
                 else if (params$genome=="rat") here("shared_data/rat_mart.RDS"))


gene_info_list<- getBM(filters= case_when(params$genes=="NCBI" ~"entrezgene_accession",
                                          params$genes=="ENSEMBL" ~ "ensembl_gene_id"), 
                       attributes= c("entrezgene_accession", "ensembl_gene_id","description"),
                       values=rownames(y_all),
                       mart=mart) %>% 
  dplyr::rename("Name"=entrezgene_accession, 
                "ENSEMBL_ID"=ensembl_gene_id)

write_tsv(x = log_cpm_df_all, file=paste0(data_dir, "log_CPM_values_by_sample.tsv"))
```

## Session info

```{r}
sessionInfo()
```

# Assesment of data quality {.tabset}
```{r}
n_samples <- nrow(combined_pheno_data_single_tissue)
n_genes <-nrow(y_all)
n_counts_per_sample <- colSums(rna_counts)
n_counts_total <-  sum(n_counts_per_sample)
n_counts_per_gene_per_sample <- (n_counts_total/n_genes)/n_samples
```

From the sequencing, we obtain `r I(n_genes)` genes that are expressed in a appreciable amount across the samples to be used for our downstream analyses. With a final, filtered transcript count of `r I(round(n_counts_total/1E+06))` million across the `r I(n_samples)` samples, we have an average of `r I(round(n_counts_per_gene_per_sample))` RNA-transcripts for each gene in each sample.

Check out the tabs below to see various QC-visualizations of the data:


## MDS 2D-plot {.tabset}

The figures below show the location of the samples in reduced 2-dimensional space.

### Dimensions 1 & 2 {.tabset}
```{r, results='asis', warning=FALSE, message=FALSE}
meta_variables <- c("BMI","smoker","SEX","AGE","Glu_tol_3_class","FAMILY_NR_new", "sourceId" )
mds_plots <- ggplot_mds_repel_updated(y = y_all, dims = c(1,2), color_by = meta_variables) %>% 
  set_names(meta_variables)
```


#### Colored by BMI
```{r, warning=FALSE}
mds_plots$BMI
```

#### Colored by smoker
```{r, warning=FALSE}
mds_plots$smoker
```
#### Colored by SEX
```{r, warning=FALSE}
mds_plots$SEX
```

#### Colored by AGE
```{r, warning=FALSE}
mds_plots$AGE
```

#### Colored by Glu_tol_3_class
```{r, warning=FALSE}
mds_plots$Glu_tol_3_class
```

#### Colored by FAMILY_NR_new
```{r, warning=FALSE}
mds_plots$FAMILY_NR_new
```

#### Colored by sourceId
```{r, warning=FALSE}
mds_plots$sourceId
```
### Dimensions 3 & 4 {.tabset}

```{r, results="asis"}
mds_plots_dim_3_4 <- ggplot_mds_repel_updated(y = y_all, dims = c(3,4), color_by = meta_variables) %>% 
  set_names(meta_variables)
```
#### Colored by BMI
```{r, warning=FALSE}
mds_plots_dim_3_4$BMI
```

#### Colored by smoker
```{r, warning=FALSE}
mds_plots_dim_3_4$smoker
```
#### Colored by SEX
```{r, warning=FALSE}
mds_plots_dim_3_4$SEX
```

#### Colored by AGE
```{r, warning=FALSE}
mds_plots_dim_3_4$AGE
```

#### Colored by Glu_tol_3_class
```{r, warning=FALSE}
mds_plots_dim_3_4$Glu_tol_3_class
```

#### Colored by FAMILY_NR_new
```{r, warning=FALSE}
mds_plots_dim_3_4$FAMILY_NR_new
```

#### Colored by sourceId
```{r, warning=FALSE}
mds_plots_dim_3_4$sourceId
```


## MA plots
The plots below show the log2-fold change in gene expression for all genes in each sample, compared to the rest of the samples.

```{r}
ggplot_MD_updated(y_all, ncol=4)
```

## Raw gene counts
Table of 100 most outlying genes (measured in terms of log2FC in a sample compared to other samples), and the corresponding transcript counts within the samples.
```{r}
MD_plot_data <- get_MD_data_for_all_samples_updated(y_all)
top_100_diff_genes <- MD_plot_data %>% 
  arrange(desc(Diff)) %>% 
  distinct(Gene, .keep_all = T) %>% 
  head(100)

y_all[top_100_diff_genes$Gene, ]$counts %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene") %>% 
  left_join(top_100_diff_genes[c("Sample", "Gene")], by=c("Gene"="Gene")) %>% 
  relocate(Sample, .after = Gene) %>% 
  mutate(across(where(is.numeric), round)) %>% 
  DT::datatable(options = list(scrollX = TRUE))
```


# Analysis Code {.tabset}

## EdgeR all

```{r}
colnames(design_all) <- make.names(colnames(design_all))

# contrast_matrix_all <- makeContrasts(BMI=BMI,
# smokerYes=smoker2,
# SEXFemale=SEX2,
# Age=AGE,
# HOMA_B_2.88=HOMA_B_2.88,
# Insulin_Resistance=HOMA_IR_156,
# levels=design_all)

fit_all <- glmQLFit(y_all, design = design_all, robust = TRUE)

# contrasts <- colnames(contrast_matrix_all)
coefficients <- design_all %>% 
  colnames() %>% 
  `[`(-1) %>% 
  set_names()

DE_genes_test_all <- coefficients %>% 
  purrr::map(~edgeR_tester(coef = .x, 
                           efit = fit_all, 
                           id_name = "Gene"))

DE_genes_test_all_w_info <- DE_genes_test_all %>% 
  map(~dplyr::left_join(., y=gene_info_list, 
                        by= if(params$genes=="NCBI") c("Gene"="Name") 
                        else c("Gene"="ENSEMBL_ID"),
                        all.x=T))
```

## GO all

```{r, message=FALSE, class.source = 'fold-show'}
GO_terms_all <- 
  get_enrichment_terms(org_db = 
                         if (params$genome == "human") {org.Hs.eg.db
                         } else if (params$genome=="mouse") {org.Mm.eg.db
                         } else if (params$genome=="rat") {org.Rn.eg.db},
                       gene_ids = rownames(y_all), gene_id_key_type = "SYMBOL",
                       min_genes = 5, max_genes = 500, )$BP

ontology_tests_all <- run_ontology_tests(GO_terms_all, 
                                         y_all, coefs = coefficients,
                                         fun=limma::camera)

# COPY STUFF FROM 0193 here
```


# Results

## Summary
We modeled the gene expression in the samples using only the knockdown or control treatment. Below, you can see a small overview of the results.

```{r, warning=F}
coefficient_count <- 1
summary_table <- map(list("Differentially expressed genes (FDR < 5%)"=DE_genes_test_all_w_info, 
                          "Enriched GO-terms (FDR < 5%)"=ontology_tests_all), 
                     n_sig_genes_pr_contrast) %>% 
  bind_rows(.id="Analysis") %>% knitr::kable()
summary_table
```

## `r coefficients[coefficient_count]` {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable(type = "edgeR")
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[[coefficient_count]] %>% 
  get_DE_datatable(type = "edgeR")
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[[coefficient_count]] %>%
  ggplot_volcano_updated(type="edgeR")
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[[coefficient_count]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[[coefficient_count]] %>%
  get_GO_datatable()
```

## smokerYes {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[["smokerYes"]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable(type = "edgeR")
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[["smokerYes"]] %>% 
  get_DE_datatable(type = "edgeR")
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[["smokerYes"]] %>%
  ggplot_volcano_updated(type="edgeR")
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[["smokerYes"]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[["smokerYes"]] %>%
  get_GO_datatable()
```

## SEXFemale {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[["SEXFemale"]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable(type = "edgeR")
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[["SEXFemale"]] %>% 
  get_DE_datatable(type = "edgeR")
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[["SEXFemale"]] %>%
  ggplot_volcano_updated(type="edgeR")
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[["SEXFemale"]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[["SEXFemale"]] %>%
  get_GO_datatable()
```

## Beta_Cell_Function {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[["HOMA_B_2.88"]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable(type = "edgeR")
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[["HOMA_B_2.88"]] %>% 
  get_DE_datatable(type = "edgeR")
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[["HOMA_B_2.88"]] %>%
  ggplot_volcano_updated(type="edgeR")
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[["HOMA_B_2.88"]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[["HOMA_B_2.88"]] %>%
  get_GO_datatable()
```

## Insulin_Resistance {.tabset}
### Differential gene expression analysis {.tabset}
#### Significant genes (FDR<0.05)
```{r, cache=FALSE, warning=FALSE}
DE_genes_test_all_w_info[["HOMA_IR_156"]] %>% 
  dplyr::filter(FDR<0.05) %>% 
  get_DE_datatable(type = "edgeR")
```

#### All genes (FDR<0.05)
```{r, warning=FALSE}
DE_genes_test_all_w_info[["HOMA_IR_156"]] %>% 
  get_DE_datatable(type = "edgeR")
```
#### Volcano plot
```{r}
DE_genes_test_all_w_info[["HOMA_IR_156"]] %>%
  ggplot_volcano_updated(type="edgeR")
```

### GO-term analysis {.tabset}
#### Significant GO-terms (FDR<0.5)
```{r}
ontology_tests_all[["HOMA_IR_156"]] %>% 
  filter(FDR<0.05) %>%
  get_GO_datatable()
```
#### All GO-terms
```{r, warning=F}
ontology_tests_all[["HOMA_IR_156"]] %>%
  get_GO_datatable()
```


