---
title: "Transcriptome analysis of the `r params$tissue`-biopsies from the QTL family cohort"
author: 
- "Analysis by Thomas Gade Koefoed"
- "Research assistant, Hansen Group"
- "Novo Nordisk Foundation Center For Basic Metabolic Research"
- "thomas.koefoed@sund.ku.dk"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
   html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
    df_print: kable
params:
  genome: human
  genes: NCBI
  tissue: fat
  exclude: none
  report_name: `r stringr::str_glue("{date}_{script_name}_{params$tissue}")`
editor_options:
  chunk_output_type: inline
---
# Pre-processing {.tabset}
## Setup
```{r setup, warning=FALSE, message=FALSE, results="hide"}
library("tidyverse") # general data manipulation

script_name <- rstudioapi::getSourceEditorContext() %>% 
  `[[`("path") %>% 
  basename() %>% 
  tools::file_path_sans_ext()

date <- format(Sys.time(), "%Y-%m-%d-%H.%M")

data_dir <- stringr::str_glue("{params$report_name}_data/")
dir.create(data_dir, showWarnings = F)

suppressPackageStartupMessages({
  # library(conflicted)
  library('biomaRt') # downloading gene data
  library("DT") # for pretty printing of paged tables
  library("edgeR") # differential expression analysis
  library("here") # Better structure and relative paths
  library("GO.db") # downloading gene data
  
  if(params$genome == "human") {library("org.Hs.eg.db")
  } else if (params$genome == "mouse") {library("org.Mm.eg.db")
  } else if (params$genome == "rat") {library("org.Rn.eg.db")}
  
  library("AnnotationDbi")
  library("ggrepel") 
  library("here") # better path management
  library("cbmr") # in-house convenience functions
  library("plotly") # interactive plots
  # library("familyStudyPackage") # own package
})
i_am("DESCRIPTION")
devtools::load_all(here())
theme_set(theme_bw())
```

## Project specific data loading and formattings {.tabset}
### Create metadata framework from Salmon data
```{r}
combined_pheno_data <- get_combined_pheno_data()
combined_pheno_data_single_tissue <- combined_pheno_data %>% 
  filter(Tissue_fixed == params$tissue) %>% 
  filter(sourceId!="641-9")
```

## DGE list creation for all samples

```{r, message = FALSE, class.source = 'fold-show'}
rna_counts_unordered <- get_rna_counts()
rna_counts <- rna_counts_unordered[,as.character(combined_pheno_data_single_tissue$Salmon.ID.raw)]
y_all <- DGEList(rna_counts, 
                 samples=combined_pheno_data_single_tissue)

log_cpm_all <- cpm(y_all, log=T) 
log_cpm_df_all <- log_cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

cpm_all <- cpm(y_all) 
cpm_df_all <- cpm_all %>% as.data.frame() %>%  
  rownames_to_column("Name")

design_all <-  model.matrix(~SEX + AGE + HOMA_IR_156, 
                            data=y_all$samples, 
                            na.action="na.pass")

idx_expressed_all <- filterByExpr(y_all, design = design_all) 


y_all <- y_all[idx_expressed_all,] %>% 
  calcNormFactors() %>%
  estimateDisp(design=design_all)
```

## Session info
```{r}
sessionInfo()
```

# Assesment of data quality {.tabset}
```{r}
n_samples <- nrow(combined_pheno_data_single_tissue)
n_genes <-nrow(y_all)
n_counts_per_sample <- colSums(rna_counts)
n_counts_total <-  sum(n_counts_per_sample)
n_counts_per_gene_per_sample <- (n_counts_total/n_genes)/n_samples
```

From the sequencing, we obtain `r I(n_genes)` genes that are expressed in a appreciable amount across the samples to be used for our downstream analyses. With a final, filtered transcript count of `r I(round(n_counts_total/1E+06))` million across the `r I(n_samples)` samples, we have an average of `r I(round(n_counts_per_gene_per_sample))` RNA-transcripts for each gene in each sample.

Check out the tabs below to see various QC-visualizations of the data:


## MDS 2D-plot {.tabset}

The figures below show the location of the samples in reduced 2-dimensional space.

### Dimensions 1 & 2 {.tabset}
```{r, results='asis', warning=FALSE, message=FALSE}
meta_variables <- c("BMI","smoker","SEX","AGE","Glu_tol_3_class","FAMILY_NR_new", "sourceId" )
mds_plots <- ggplot_mds_repel_updated(y = y_all, dims = c(1,2), color_by = meta_variables) %>% 
  set_names(meta_variables)
```


#### Colored by BMI
```{r, warning=FALSE}
mds_plots$BMI
```

#### Colored by smoker
```{r, warning=FALSE}
mds_plots$smoker
```
#### Colored by SEX
```{r, warning=FALSE}
mds_plots$SEX
```

#### Colored by AGE
```{r, warning=FALSE}
mds_plots$AGE
```

#### Colored by Glu_tol_3_class
```{r, warning=FALSE}
mds_plots$Glu_tol_3_class
```

#### Colored by FAMILY_NR_new
```{r, warning=FALSE}
mds_plots$FAMILY_NR_new
```

#### Colored by sourceId
```{r, warning=FALSE}
mds_plots$sourceId
```
### Dimensions 3 & 4 {.tabset}

```{r, results="asis"}
mds_plots_dim_3_4 <- ggplot_mds_repel_updated(y = y_all, dims = c(3,4), color_by = meta_variables) %>% 
  set_names(meta_variables)
```
#### Colored by BMI
```{r, warning=FALSE}
mds_plots_dim_3_4$BMI
```

#### Colored by smoker
```{r, warning=FALSE}
mds_plots_dim_3_4$smoker
```
#### Colored by SEX
```{r, warning=FALSE}
mds_plots_dim_3_4$SEX
```

#### Colored by AGE
```{r, warning=FALSE}
mds_plots_dim_3_4$AGE
```

#### Colored by Glu_tol_3_class
```{r, warning=FALSE}
mds_plots_dim_3_4$Glu_tol_3_class
```

#### Colored by FAMILY_NR_new
```{r, warning=FALSE}
mds_plots_dim_3_4$FAMILY_NR_new
```

#### Colored by sourceId
```{r, warning=FALSE}
mds_plots_dim_3_4$sourceId
```


## MA plots
The plots below show the log2-fold change in gene expression for all genes in each sample, compared to the rest of the samples.

```{r, fig.height=30}
ggplot_MD_updated(y_all, ncol=4)
```

## Raw gene counts
Table of 100 most outlying genes (measured in terms of log2FC in a sample compared to other samples), and the corresponding transcript counts within the samples.
```{r}
MD_plot_data <- get_MD_data_for_all_samples_updated(y_all)
top_100_diff_genes <- MD_plot_data %>% 
  arrange(desc(Diff)) %>% 
  distinct(Gene, .keep_all = T) %>% 
  head(100)

y_all[top_100_diff_genes$Gene, ]$counts %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene") %>% 
  left_join(top_100_diff_genes[c("Sample", "Gene")], by=c("Gene"="Gene")) %>% 
  relocate(Sample, .after = Gene) %>% 
  mutate(across(where(is.numeric), round)) %>% 
  DT::datatable(options = list(scrollX = TRUE))
```
